# **WORDNET & NLP**

Topics: WordNet Basics  
- Synsets  
- Lemmas  
- Hypernyms  
- Similarity Metrics  
- NLP Applications

## **1. WHAT IS WORDNET?**

**WordNet** is a large lexical database (dictionary) of English, developed at Princeton University. It groups English words into sets of synonyms called **synsets**, and records various semantic and lexical relationships between them.


| **🎯  Think of it this way:** |
| - |
| A normal dictionary tells you the DEFINITION of a word. |
| WordNet tells you HOW words RELATE to each other. |
| Example: It knows that 'dog' is a type of 'animal', that 'poodle' is a type of 'dog', |
| and that 'dog' is the opposite of 'cat' in some contexts. |


### **1.1) Why is WordNet Important for NLP?**

WordNet is one of the most widely used resources in Natural Language Processing. It helps computers understand:

- **Word Meaning:** 'bank' as a financial institution vs. 'bank' as a riverbank
- **Word Relationships:** 'car' is similar to 'vehicle', 'automobile', 'motorcar'
- **Semantic Similarity:** How similar are the words 'doctor' and 'nurse'?
- **Text Classification:** Grouping news articles by topic
- **Question Answering:** Understanding that 'canine' = 'dog'

### **1.2) WordNet as an NLTK Corpus**

In Python, WordNet is accessed through the NLTK (Natural Language Toolkit) library. It is treated as a corpus reader — a structured data source that NLTK can query.

Two ways to import WordNet:

```python
from nltk.corpus import wordnet         \# Full name
from nltk.corpus import wordnet as wn   \# Short alias (recommended)
```

## **2. INSTALLATION & SETUP**


### **Step 1: Install NLTK**

```python
pip install nltk
```


### **Step 2: Download WordNet Data**

```python
import nltk
nltk.download('wordnet')       # Main WordNet database
nltk.download('omw-1.4')       # Open Multilingual WordNet (for other languages)
nltk.download('popular')       # for popular databases
nltk.download('wordnet\_ic')   # Information Content files (for similarity)
```

### **Step 3: Verify Installation**


In [2]:
from nltk.corpus import wordnet as wn
display(wn.synsets('dog')) 

[Synset('dog.n.01'),
 Synset('frump.n.01'),
 Synset('dog.n.03'),
 Synset('cad.n.01'),
 Synset('frank.n.02'),
 Synset('pawl.n.01'),
 Synset('andiron.n.01'),
 Synset('chase.v.01')]

## **3. CORE TERMINOLOGY — DEEP DIVE**

Before writing any code, you must understand these core terms. Every WordNet operation revolves around them.
| **TERM** | **MEANING** |
| :-: | - |
| **WORD** | A regular English word like 'dog', 'run', 'happy'. A single word can have multiple meanings. |
| **SENSE** | One specific meaning of a word. 'Bank' has at least 2 senses: <br>(1) financial institution, <br>(2) riverbank. |
| **SYNSET** | A Synonym Set — a group of words that share the same meaning. E.g., \{car, auto, automobile, motorcar\} is one synset. |
| **LEMMA** | A specific word inside a specific synset. It represents one word-sense pairing. E.g., Lemma('dog.n.01.domestic\_dog') |
| **POS** | Part of Speech — whether a word is a Noun (`n`), Verb (`v`), Adjective (`a/s`), or Adverb (`r`). |
| **HYPERNYM** | The '`is-a`' parent. Animal is a hypernym of Dog **\`Dog is a animal\`**. |
| **HYPONYM** | The '`is-a`' child. Dog is a hyponym of Animal **\`Dog is a animal\`.** |
| **HOLONYM** | The '`part-of`' relationship. 'pack' is a holonym of 'dog' **\`dogs are part of a pack\`**. |
| **MERONYM** | The '`has-part`' relationship. 'dog' is a meronym of 'pack'. |
| **ANTONYM** | The opposite word. 'good' and 'bad' are antonyms. |



### **3.1) The Synset ID Format**

Every synset has a 3-part identifier: word.pos.nn


| **Part** | **Meaning** | **Example** |
| :-: | :-: | :-: |
| word | The base word | dog |
| pos | n=noun, v=verb, a=adj, r=adverb | n |
| nn | Sense number (01 = most common) | 01 |


So dog.n.01 means: the word 'dog', as a noun, first/most common sense.

*Example:*



In [ ]:
for i in wn.synsets('quote'):
    print(f"{str(i)}, \nDef: {str(i.definition())}, \nExample: {i.examples()}\nPOS: {i.pos()}\n------")

Synset('quotation_mark.n.01'), 
Def: a punctuation mark used to attribute the enclosed text to someone else, 
Example: []
POS: n
------
Synset('quotation.n.02'), 
Def: a passage or expression that is quoted or cited, 
Example: []
POS: n
------
Synset('quote.v.01'), 
Def: repeat a passage from, 
Example: ['He quoted the Bible to her']
POS: v
------
Synset('quote.v.02'), 
Def: name the price of, 
Example: ['quote prices for cars']
POS: v
------
Synset('quote.v.03'), 
Def: refer to for illustration or proof, 
Example: ['He said he could quote several instances of this behavior']
POS: v
------
Synset('quote.v.04'), 
Def: put quote marks around, 
Example: ['Here the author is quoting his colleague']
POS: v
------


## **4. SYNSETS — SYNONYM SETS**

**A Synset (Synonym Set)** is the fundamental building block of WordNet. It is a group of words that all share the same meaning in a specific context.


### **4.1) Getting Synsets for a Word**



In [7]:
from nltk.corpus import wordnet as wn
synsets = wn.synsets('dog')
display(synsets)

[Synset('dog.n.01'),
 Synset('frump.n.01'),
 Synset('dog.n.03'),
 Synset('cad.n.01'),
 Synset('frank.n.02'),
 Synset('pawl.n.01'),
 Synset('andiron.n.01'),
 Synset('chase.v.01')]

Notice that 'dog' has **8 different synsets** — meaning 8 different senses/meanings!

### **4.2) Filtering by Part of Speech (POS)**


In [9]:
# Get only NOUN synsets
display(wn.synsets('dog', pos=wn.NOUN)) 

[Synset('dog.n.01'),
 Synset('frump.n.01'),
 Synset('dog.n.03'),
 Synset('cad.n.01'),
 Synset('frank.n.02'),
 Synset('pawl.n.01'),
 Synset('andiron.n.01')]

In [10]:
# Get only VERB synsets
display(wn.synsets('dog', pos=wn.VERB)) 

[Synset('chase.v.01')]

In [11]:
# Get only VERB synsets
display(wn.synsets('dog', pos=wn.ADJ)) 

[]

In [12]:
# Get only VERB synsets
display(wn.synsets('dog', pos=wn.ADV)) 

[]

In [13]:
# Get only VERB synsets
display(wn.synsets('dog', pos='s')) # 's' = satellite adjective (a type of adjective)

[]


### **4.3) Getting Info from a Synset**

In [15]:
# Access a specific synset directly

dog_synset = wn.synset('dog.n.01')
print(dog_synset)

Synset('dog.n.01')


In [17]:
# 1. Get the DEFINITION
print("Definition:", dog_synset.definition())

# 2. Get USAGE EXAMPLES
print("Examples:", dog_synset.examples())

# 3. Get all LEMMAS (words in this synset)
print("Lemmas:", dog_synset.lemmas())

# 4. Get just the lemma NAMES
print("Lemma Names:", [lemma.name() for lemma in dog_synset.lemmas()])

# 5. Get the POS tag
print("POS Tag:", dog_synset.pos())

# 6. Get the synset NAME
print("Synset Name:", dog_synset.name())

Definition: a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds
Examples: ['the dog barked all night']
Lemmas: [Lemma('dog.n.01.dog'), Lemma('dog.n.01.domestic_dog'), Lemma('dog.n.01.Canis_familiaris')]
Lemma Names: ['dog', 'domestic_dog', 'Canis_familiaris']
POS Tag: n
Synset Name: dog.n.01



### **4.4) Getting Synonyms (All Words in a Synset Group)**


In [22]:
# wn.synonyms() returns ALL synonym groups for a word
display("Synonyms:", wn.synonyms('car'))

'Synonyms:'

[['auto', 'automobile', 'machine', 'motorcar'],
 ['railcar', 'railroad_car', 'railway_car'],
 ['gondola'],
 ['elevator_car'],
 ['cable_car']]

| **Key Insight:** |
| - |
| Different senses of 'car' are NOT synonyms of each other! |
|1. An 'automobile' and a 'railcar' are both called 'car' but they are different things. |
|2. That is why wn.synonyms() returns a NESTED list — each inner list is one meaning. |
|3. Words within the same inner list ARE true synonyms. |

## **5. LEMMAS — WORD FORMS**

**A Lemma** in WordNet is a specific word paired with a specific synset (meaning). While a Synset represents a concept, a Lemma represents a WORD that expresses that concept.


| **Synset vs Lemma — The Key Difference** |
| - |
| SYNSET = the CONCEPT  (e.g., the idea of a domesticated dog) |
| LEMMA  = a WORD for that concept  (e.g., 'dog', 'domestic\_dog', 'Canis\_familiaris') |
| One synset can have MULTIPLE lemmas (synonyms). |
| One lemma belongs to EXACTLY ONE synset. |

### **5.1) Accessing Lemmas**


In [24]:
from nltk.corpus import wordnet as wn
# Get lemma directly using word.pos.sensenum.lemma_name format

lemma = wn.lemma('dog.n.01.dog')
print(lemma)           # Lemma('dog.n.01.dog')
print("Name:", lemma.name())  # dog
print("Synset:", lemma.synset())  # Synset('dog.n.01')
print("Key:", lemma.key())     # Unique sense key string
print("Count:", lemma.count())   # How often used in SemCor corpus

Lemma('dog.n.01.dog')
Name: dog
Synset: Synset('dog.n.01')
Key: dog%1:05:00::
Count: 42


### **5.2) Lemma Relationships**

Some relations in WordNet only exist at the LEMMA level, not the synset level:

In [43]:
# ANTONYMS — opposite words (only on lemmas!)
good = wn.lemma('good.a.01.good')
print("Antonyms:", good.antonyms()[0].name())

# DERIVATIONALLY RELATED FORMS
vocal = wn.lemma('vocal.a.01.vocal')
print("Derivationally Related Forms:", vocal.derivationally_related_forms()[0])

# PERTAINYMS — adjective derived from a noun
print("Pertainyms:", vocal.pertainyms()[0])

Antonyms: bad
Derivationally Related Forms: Lemma('vocalize.v.02.vocalize')
Pertainyms: Lemma('voice.n.02.voice')


| **Relation** | **Description** | **Example** |
| - | - | - |
| `antonyms()` | Opposite in meaning | good → bad |
| `derivationally_related_forms()` | Different POS from same root | vocal → vocalize |
| `pertainyms()` | Adj derived from a noun | vocal → voice |

## **6. LEXICAL RELATIONS**


WordNet organizes words in a rich hierarchy of relationships. Understanding these is critical for NLP tasks.


### **6.1) The WordNet Hierarchy**

The most important structure is the **IS-A hierarchy** (also called the hypernym/hyponym tree). It works exactly like a taxonomy in biology:

### **6.2) Hypernyms — Going UP the Tree**
**Hypernym** = the more GENERAL category. 'Dog' is a type of 'canine'. 'Canine' is the hypernym of 'Dog'.

In [47]:
dog = wn.synset('dog.n.01')

# Immediate parent(s)
print("Hypernyms of 'dog':", dog.hypernyms())   

# All ancestors up to root
hyper = lambda s: s.hypernyms()
ancestors = list(dog.closure(hyper))
print("Ancestors of 'dog':", sorted(ancestors))

# Get the very top (root) of the tree
print("Root hypernyms of 'dog':", dog.root_hypernyms())

Hypernyms of 'dog': [Synset('canine.n.02'), Synset('domestic_animal.n.01')]
Ancestors of 'dog': [Synset('animal.n.01'), Synset('canine.n.02'), Synset('carnivore.n.01'), Synset('chordate.n.01'), Synset('domestic_animal.n.01'), Synset('entity.n.01'), Synset('living_thing.n.01'), Synset('mammal.n.01'), Synset('object.n.01'), Synset('organism.n.01'), Synset('physical_entity.n.01'), Synset('placental.n.01'), Synset('vertebrate.n.01'), Synset('whole.n.02')]
Root hypernyms of 'dog': [Synset('entity.n.01')]


### **6.3) Hyponyms — Going DOWN the Tree**
**Hyponym** = a more SPECIFIC category. 'Poodle' is a type of 'dog'. 'Poodle' is a hyponym of 'Dog'.

In [51]:
# Immediate children
display("Direct hyponyms (children): (top 5)", sorted(dog.hyponyms())[:5])  # First 5 only

# All descendants (transitive closure)
hypo = lambda s: s.hyponyms()
all_dogs = sorted(dog.closure(hypo))[:5]
display("All descendants:", all_dogs)

'Direct hyponyms (children): (top 5)'

[Synset('basenji.n.01'),
 Synset('corgi.n.01'),
 Synset('cur.n.01'),
 Synset('dalmatian.n.02'),
 Synset('great_pyrenees.n.01')]

'All descendants:'

[Synset('affenpinscher.n.01'),
 Synset('afghan_hound.n.01'),
 Synset('airedale.n.01'),
 Synset('american_foxhound.n.01'),
 Synset('american_staffordshire_terrier.n.01')]

### **6.4) Finding the Common Ancestor**

**Lowest Common Hypernym (LCH)** = the most specific concept that two words share as an ancestor. This is used to measure semantic similarity.

In [54]:
dog = wn.synset('dog.n.01')
cat = wn.synset('cat.n.01')
woman = wn.synset('woman.n.01')
man = wn.synset('man.n.01')

# What do dog and cat have in common?
print("Lowest common hypernym of 'dog' and 'cat':", dog.lowest_common_hypernyms(cat)[0])

# What do dog and woman have in common?
print("Lowest common hypernym of 'dog' and 'woman':", man.lowest_common_hypernyms(woman)[0])

Lowest common hypernym of 'dog' and 'cat': Synset('carnivore.n.01')
Lowest common hypernym of 'dog' and 'woman': Synset('adult.n.01')


### **6.5) Holonyms and Meronyms — Part/Whole Relations**
**Holonym** is the entire object (e.g., "car"), while a **meronym** is a constituent part of that object (e.g., "wheel").

In [59]:
dog = wn.synset('dog.n.01')

# MEMBER holonyms: 'dog' is a member of what?
print("member holonyms of dog are: ", dog.member_holonyms())   # dog is a member of a 'pack'

# PART meronyms: what are the parts of 'dog'?
body = wn.synset('body.n.01')
print("meronym of body is: ", body.part_meronyms()[:3])   # body has parts: arm, leg, head...

# SUBSTANCE meronyms: made of what?
wood = wn.synset('wood.n.01')
print("substance meronyms of wood are: ", wood.substance_meronyms())

member holonyms of dog are:  [Synset('canis.n.01'), Synset('pack.n.06')]
meronym of body is:  [Synset('arm.n.01'), Synset('vascular_system.n.01'), Synset('crotch.n.02')]
substance meronyms of wood are:  [Synset('lignin.n.01')]


| **Method** | **Meaning** | **Example** |
| - | - | - |
| hypernyms() | More general (parent) | dog → canine → animal |
| hyponyms() | More specific (child) | dog → poodle, corgi |
| root\_hypernyms() | Top of the hierarchy | dog → entity.n.01 |
| lowest\_common\_hypernyms() | Shared ancestor | dog+cat → carnivore |
| member\_holonyms() | X is member of what? | dog → pack |
| member\_meronyms() | What are members of X? | pack → dog |
| part\_holonyms() | X is part of what? | finger → hand |
| part\_meronyms() | What are parts of X? | hand → finger, palm |
| substance\_holonyms() | X is substance of what? | H2O → water |
| substance\_meronyms() | Made of what substance? | wood → lignin |
| closure(relation) | All transitive relations | all ancestors of dog |

## **7. SIMILARITY METRICS**
WordNet can calculate how **semantically similar** two words are. This is extremely useful for NLP tasks like finding synonyms, topic modeling, and building chatbots.


| **Setup for all similarity examples:** |
| - |
| from nltk.corpus import wordnet as wn |
| from nltk.corpus import wordnet\_ic |
| dog = wn.synset('dog.n.01') |
| cat = wn.synset('cat.n.01') |
| hit = wn.synset('hit.v.01') |
| slap = wn.synset('slap.v.01') |
| brown\_ic = wordnet\_ic.ic('ic-brown.dat')   \# Information Content corpus |
| semcor\_ic = wordnet\_ic.ic('ic-semcor.dat') \# Another IC corpus |


### **7.1) Path Similarity**
**`path_similarity()`** counts how many steps through the hypernym tree separate two words. `Shorter path = more similar`.

*Formula:* 
$$\text{score} = \frac{1}{\text{shortest\_path\_length} + 1}$$
→  Range: $0$ (unrelated) to $1$ (identical)

In [73]:
dog = wn.synset('dog.n.01')
cat = wn.synset('cat.n.01')
hit = wn.synset('hit.v.01')
slap = wn.synset('slap.v.01')

print("Path similarity between dog and cat:", dog.path_similarity(cat))     # Similar — both mammals
print("Path similarity between hit and slap:", round(hit.path_similarity(slap),3))    # Less similar — both actions but different

# You can also call it on the wn module directly:
print("Path similarity between hit and slap (direct call):", round(wn.path_similarity(hit, slap),3))

# Disable fake root node (may return None for verbs):
print("Path similarity between hit and slap (no fake root):", hit.path_similarity(slap, simulate_root=False))

Path similarity between dog and cat: 0.2
Path similarity between hit and slap: 0.143
Path similarity between hit and slap (direct call): 0.143
Path similarity between hit and slap (no fake root): None


### **6.2) Leacock-Chodorow (LCH) Similarity**

**lch\_similarity()** improves on path similarity by accounting for the DEPTH of the taxonomy. A path in a deep taxonomy is more meaningful than the same path in a shallow one.

*Formula:* 
$$\text{score} = -\ln\big(\frac{\text{path\_length}}{2 \times \text{taxonomy\_depth}}\big)$$

In [72]:
print(round(dog.lch_similarity(cat),3))   # Higher = more similar
print(round(hit.lch_similarity(slap),3))

2.028
1.312


### **6.3) Wu-Palmer (WUP) Similarity**

**wup_similarity()** uses the DEPTH of both synsets and their Least Common Subsumer (LCS) in the hierarchy. Very popular in practice.

*Formula:* 
$$\text{score} = \frac{2 \times \text{depth(LCS)}}{\text{depth(s1)} + \text{depth(s2)}}$$


In [76]:
print(round(hit.wup_similarity(slap), 3))
print(round(dog.wup_similarity(cat), 3))

0.25
0.857


### **6.4) Information Content (IC) Based Metrics**

These metrics use **Information Content (IC)** — a measure of how rare/specific a word is in a large corpus. Rare, specific words carry more information. You need to load an IC file first.

#### **6.4.1) Resnik Similarity**

Based only on the IC of the LCS (common ancestor). 
$$\text{score} = \text{IC}(\text{LCS})$$

In [83]:
from nltk.corpus import wordnet_ic
brown_ic = wordnet_ic.ic('ic-brown.dat')
semcor_ic = wordnet_ic.ic('ic-semcor.dat')

print(round(dog.res_similarity(cat, brown_ic), 3))  
print(round(dog.res_similarity(cat, semcor_ic), 3))  # Different corpus = different score!

7.912
7.255



#### **6.4.2) Jiang-Conrath (JCN) Similarity**

Considers IC of both words AND their LCS. 

$$\text{score} = \frac{1}{(\text{IC(s1)} + \text{IC(s2)} - 2 \times \text{IC(LCS)})}$$

In [90]:
print(round(dog.jcn_similarity(cat, brown_ic), 3))

0.45


#### **6.4.3) Lin Similarity**

Normalized version of Resnik. 

$$\text{score} = \frac{2 \times \text{IC(LCS)}}{(\text{IC(s1)} + \text{IC(s2)})}$$
→  Range: 0 to 1

In [92]:
print(round(dog.lin_similarity(cat, semcor_ic),3))

0.886


89% similar — dog and cat are very related concepts!

#### **Comparison of All Similarity Metrics**

| **Metric** | **Based On** | **Range** | **Needs IC?** | **dog vs cat** |
| - | - | - | - | - |
| path\_similarity | Shortest path | 0 to 1 | No | 0.2 |
| lch\_similarity | Path + depth | \>0 (log scale) | No | 2.028 |
| wup\_similarity | Depths of synsets | 0 to 1 | No | 0.857 |
| res\_similarity | IC of LCS | 0 to ∞ | Yes | 7.911 |
| jcn\_similarity | IC of both + LCS | 0 to ∞ | Yes | 0.449 |
| lin\_similarity | Normalized IC | 0 to 1 | Yes | 0.886 |

## **8. MORPHY — LEMMATIZATION**

**Morphy** is WordNet's built-in lemmatizer. It converts inflected word forms back to their base (lemma) form. For example, 'dogs' → 'dog', 'running' → 'run', 'better' → 'good'.


| **Why do we need Morphy?** |
| - |
| WordNet stores only BASE FORMS. So if you search for 'dogs', you won't find it directly. |
| Morphy converts 'dogs' → 'dog', then WordNet can find the synsets. |
| This is called LEMMATIZATION — a key NLP preprocessing step. |


### **8.1) Using wn.morphy()**


In [95]:
from nltk.corpus import wordnet as wn

# Basic lemmatization
print(wn.morphy('dogs'))         # plural noun → base
print(wn.morphy('churches'))     # irregular plural
print(wn.morphy('aardwolves'))   # very irregular plural
print(wn.morphy('abaci'))        # Latin plural → abacus


# With POS context
print(wn.morphy('denied', wn.VERB))   # verb inflection
print(wn.morphy('denied', wn.NOUN))   # None — 'denied' is not a noun

# Now search using the lemmatized form
display(wn.synsets('denied', wn.VERB))  # WordNet auto-lemmatizes

dog
church
aardwolf
abacus
deny
None


[Synset('deny.v.01'),
 Synset('deny.v.02'),
 Synset('deny.v.03'),
 Synset('deny.v.04'),
 Synset('deny.v.05'),
 Synset('traverse.v.03'),
 Synset('deny.v.07')]

### **8.2 WordNetLemmatizer**

NLTK also provides a WordNetLemmatizer class that wraps Morphy for easy use:

In [97]:
from nltk.stem.wordnet import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
print("eggs =>",lemmatizer.lemmatize('eggs', pos='n'))   # noun
print("legs =>",lemmatizer.lemmatize('legs', pos='n'))   # noun
print("running =>",lemmatizer.lemmatize('running', pos='v')) # verb
print("better =>",lemmatizer.lemmatize('better', pos='a'))  # adjective

eggs => egg
legs => leg
running => run
better => good


## **9. MULTILINGUAL WORDNET**

WordNet in NLTK supports **32+ languages** through the Open Multilingual WordNet (OMW). You can look up words, get lemmas, and even find synsets in other languages!


### **9.1) Supported Languages**



In [103]:
import nltk
nltk.download('omw-1.4')  # Download multilingual data first
from nltk.corpus import wordnet as wn
print(sorted(wn.langs()))
print(len(wn.langs()), "languages available in WordNet")

['als', 'arb', 'bul', 'cat', 'cmn', 'dan', 'ell', 'eng', 'eus', 'fin', 'fra', 'glg', 'heb', 'hrv', 'ind', 'isl', 'ita', 'ita_iwn', 'jpn', 'lit', 'nld', 'nno', 'nob', 'pol', 'por', 'ron', 'slk', 'slv', 'spa', 'swe', 'tha', 'zsm']
32 languages available in WordNet


[nltk_data] Downloading package omw-1.4 to /home/naman/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


### **9.2) Multilingual Examples**

In [99]:
dog = wn.synset('dog.n.01')

# Get lemma names in different languages
print(dog.lemma_names('ita'))   # Italian
print(dog.lemma_names('por'))   # Portuguese
print(dog.lemma_names('dan'))   # Danish


# Search in Italian
print(wn.lemmas('cane', lang='ita'))

# Search synonyms in Spanish
print(wn.synonyms('coche', lang='spa'))

['Canis_familiaris', 'cane']
['cachorra', 'cachorro', 'cadela', 'cão']
['hund', 'køter', 'vovhund', 'vovse']
[Lemma('dog.n.01.cane'), Lemma('cramp.n.02.cane'), Lemma('hammer.n.01.cane'), Lemma('bad_person.n.01.cane'), Lemma('incompetent.n.01.cane')]
[['auto', 'automóvil', 'carro', 'máquina', 'turismo', 'vehículo'], ['automotor', 'vagón'], ['vagón', 'vagón_de_pasajeros']]


## **10. PRACTICAL NLP USE CASES**
Here are complete, runnable examples for common NLP tasks using WordNet.


### **Use Case 1: Word Sense Disambiguation (WSD)**

Problem: The word 'bank' can mean a financial institution or the side of a river. **WSD** picks the right sense based on context.

In [ ]:
from nltk.corpus import wordnet as wn
def get_definition(word, sense_number):
    synsets = wn.synsets(word, pos=wn.NOUN)
    if sense_number <= len(synsets):
        syn = synsets[sense_number - 1]
        return {
            'synset': syn.name(),
            'definition': syn.definition(),
            'examples': syn.examples()
        }
    return None

In [109]:
word = 'bank'
for i in range(1, 4):
    result = get_definition(word, i)
    if result:
        print(f'Sense {i}: {result["definition"]}')

Sense 1: sloping land (especially the slope beside a body of water)
Sense 2: a financial institution that accepts deposits and channels the money into lending activities
Sense 3: a long ridge or pile


### **Use Case 2: Semantic Similarity Between Sentences**

Compare two sentences by finding the maximum similarity between any word pair.

In [111]:
from nltk.corpus import wordnet as wn
import nltk
nltk.download('averaged_perceptron_tagger')
nltk.download('punkt')

def get_first_synset(word, pos='n'):
    synsets = wn.synsets(word, pos=pos)
    return synsets[0] if synsets else None

def sentence_similarity(sent1_words, sent2_words):
    """Compare each word in sent1 with each word in sent2."""
    scores = []
    for w1 in sent1_words:
        s1 = get_first_synset(w1)
        if not s1: continue
        for w2 in sent2_words:
            s2 = get_first_synset(w2)
            if not s2: continue
            sim = s1.wup_similarity(s2)
            if sim is not None:
                scores.append(sim)
    return sum(scores)/len(scores) if scores else 0

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /home/naman/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package punkt to /home/naman/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [112]:
# Example

s1 = ['dog', 'run', 'park']
s2 = ['puppy', 'jog', 'garden']
s3 = ['car', 'drive', 'highway']
print(f'S1 vs S2: {sentence_similarity(s1, s2):.3f}')  # Should be high
print(f'S1 vs S3: {sentence_similarity(s1, s3):.3f}')  # Should be lower

S1 vs S2: 0.344
S1 vs S3: 0.286


### **Use Case 3: Hypernym Path Explorer**

Useful for taxonomy-based text classification — trace a word up to its root.

In [113]:
def hypernym_path(word, pos='n'):
    """Show the full IS-A path for a word up to entity."""
    synsets = wn.synsets(word, pos=pos)
    if not synsets:
        return f'No synsets found for {word}'
    
    syn = synsets[0]  # Take most common sense
    paths = syn.hypernym_paths()
    path = paths[0]  # Take first path    

    print(f'Hypernym path for "{word}" ({syn.name()}):',)

    for i, s in enumerate(path):
        indent = '  ' * i
        print(f'{indent}→ {s.name()} ({s.definition()[:40]}...)')

In [114]:
hypernym_path('poodle')

Hypernym path for "poodle" (poodle.n.01):
→ entity.n.01 (that which is perceived or known or infe...)
  → physical_entity.n.01 (an entity that has physical existence...)
    → object.n.01 (a tangible and visible entity; an entity...)
      → whole.n.02 (an assemblage of parts that is regarded ...)
        → living_thing.n.01 (a living (or once living) entity...)
          → organism.n.01 (a living thing that has (or can develop)...)
            → animal.n.01 (a living organism characterized by volun...)
              → chordate.n.01 (any animal of the phylum Chordata having...)
                → vertebrate.n.01 (animals having a bony or cartilaginous s...)
                  → mammal.n.01 (any warm-blooded vertebrate having the s...)
                    → placental.n.01 (mammals having a placenta; all mammals e...)
                      → carnivore.n.01 (a terrestrial or aquatic flesh-eating ma...)
                        → canine.n.02 (any of various fissiped mammals with non...)
     

### **Use Case 4: Synonym Expansion for Search**
Improve search recall by expanding query words with their synonyms.

In [115]:
def expand_query(query_word, pos='n', max_synonyms=5):
    """Expand a search term with its WordNet synonyms."""
    synonyms = set()
    synonyms.add(query_word)  # include original
    for synset in wn.synsets(query_word, pos=pos):
        for lemma in synset.lemmas():
            name = lemma.name().replace('_', ' ')
            synonyms.add(name)   
    return list(synonyms)[:max_synonyms + 1]

print('Synonyms for car:', expand_query('car'))
print('Synonyms for happy:', expand_query('happy', pos='a'))
print('Synonyms for run:', expand_query('run', pos='v'))

Synonyms for car: ['cable car', 'gondola', 'railcar', 'elevator car', 'automobile', 'motorcar']
Synonyms for happy: ['well-chosen', 'happy', 'felicitous', 'glad']
Synonyms for run: ['break away', 'hunt down', 'bleed', 'hightail it', 'play', 'execute']


### **Use Case 5: Checking if Word Belongs to a Category**

Check if a word is a type of a given category using hypernym chains.


In [117]:
def is_a(word, category, pos='n'):
    """Check if 'word' is a type of 'category' using WordNet."""
    word_synsets = wn.synsets(word, pos=pos)
    cat_synsets = wn.synsets(category, pos=pos)
    if not word_synsets or not cat_synsets:
        return False

    word_syn = word_synsets[0]
    cat_syn = cat_synsets[0]

    # Get all hypernyms (ancestors) of the word
    hyper = lambda s: s.hypernyms()
    all_hypernyms = set(word_syn.closure(hyper))
    return cat_syn in all_hypernyms

In [118]:
# Test it out
print(is_a('poodle', 'dog'))     # True — poodle IS a dog
print(is_a('poodle', 'animal'))  # True — poodle IS an animal
print(is_a('car', 'animal'))     # False — car is NOT an animal
print(is_a('eagle', 'bird'))     # True — eagle IS a bird

True
True
False
True


| **Summary of NLP Applications of WordNet** |
| - |
| ✓  Word Sense Disambiguation — which meaning of a word is used? |
| ✓  Semantic Similarity — how similar are two words/sentences? |
| ✓  Synonym Expansion — broaden search queries |
| ✓  Taxonomy/Ontology — is X a type of Y? |
| ✓  Text Classification — group text by semantic categories |
| ✓  Named Entity Recognition — identify types of entities |
| ✓  Machine Translation — find corresponding concepts in other languages |
| ✓  Question Answering — match question words to answer words |
| ✓  Lemmatization — normalize words to base form for indexing |

